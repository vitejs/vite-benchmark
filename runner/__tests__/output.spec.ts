import { describe, expect, test } from 'vitest'
import {
  composeGitHubActionsSummary,
  calcMetrics,
  composeSummarized,
} from '../src/cases'

describe('data report', () => {
  test('composeSummarized', () => {
    const v = composeSummarized(results, compares, baseCases)
    Object.keys(v).forEach((key) => {
      v[key]!.forEach((item) => {
        item.metrics.fcpStat.kmeans = '-'
        item.metrics.serverStartStat.kmeans = '-'
        item.metrics.startupStat.kmeans = '-'
      })
    })
    expect(v).matchSnapshot()
  })

  test('calcMetrics', () => {
    const v = calcMetrics(results, 'perf-1', 'vitejs%2Fvite%40063d93b')
    v.fcpStat.kmeans = '-'
    v.serverStartStat.kmeans = '-'
    v.startupStat.kmeans = '-'
    expect(v).toMatchSnapshot()
  })

  test('composeGitHubActionsSummary', () => {
    let v = composeGitHubActionsSummary({
      options: {
        pullNumber: 12787,
      },
      compares,
      summarizedResult: composeSummarized(results, compares, baseCases),
    })

    v = v.replaceAll(/<td>.*<\/td>/g, '<td>-</td>')
    expect(v).matchSnapshot()
  })
})

const results = [
  {
    startup: 2761,
    serverStart: 457,
    fcp: 1632.9000000953674,
    index: 0,
    caseId: 'perf-1',
    displayName: 'vite 2.7 slow',
    uniqueKey: 'vitejs%2Fvite%40063d93b',
  },
  {
    startup: 2632,
    serverStart: 462.19,
    fcp: 1491.5,
    index: 0,
    caseId: 'perf-1',
    displayName: 'vite 2.7 slow',
    uniqueKey: 'sun0day%2Fvite%40999ad63',
  },
  {
    startup: 2644,
    serverStart: 189,
    fcp: 2069.5,
    index: 0,
    caseId: 'perf-2',
    displayName: '1000 React components',
    uniqueKey: 'vitejs%2Fvite%40063d93b',
  },
  {
    startup: 2642,
    serverStart: 197.97,
    fcp: 2058.100000143051,
    index: 0,
    caseId: 'perf-2',
    displayName: '1000 React components',
    uniqueKey: 'sun0day%2Fvite%40999ad63',
  },
  {
    startup: 2000,
    serverStart: 168,
    fcp: 1531,
    index: 1,
    caseId: 'perf-1',
    displayName: 'vite 2.7 slow',
    uniqueKey: 'vitejs%2Fvite%40063d93b',
  },
  {
    startup: 1956,
    serverStart: 153.41,
    fcp: 1565.2999999523163,
    index: 1,
    caseId: 'perf-1',
    displayName: 'vite 2.7 slow',
    uniqueKey: 'sun0day%2Fvite%40999ad63',
  },
  {
    startup: 2671,
    serverStart: 203,
    fcp: 2201.5999999046326,
    index: 1,
    caseId: 'perf-2',
    displayName: '1000 React components',
    uniqueKey: 'vitejs%2Fvite%40063d93b',
  },
  {
    startup: 2644,
    serverStart: 214.28,
    fcp: 2150.300000190735,
    index: 1,
    caseId: 'perf-2',
    displayName: '1000 React components',
    uniqueKey: 'sun0day%2Fvite%40999ad63',
  },
  {
    startup: 2038,
    serverStart: 163,
    fcp: 1634.3000001907349,
    index: 2,
    caseId: 'perf-1',
    displayName: 'vite 2.7 slow',
    uniqueKey: 'vitejs%2Fvite%40063d93b',
  },
  {
    startup: 1957,
    serverStart: 165.6,
    fcp: 1545.9000000953674,
    index: 2,
    caseId: 'perf-1',
    displayName: 'vite 2.7 slow',
    uniqueKey: 'sun0day%2Fvite%40999ad63',
  },
  {
    startup: 2584,
    serverStart: 212,
    fcp: 2088.5999999046326,
    index: 2,
    caseId: 'perf-2',
    displayName: '1000 React components',
    uniqueKey: 'vitejs%2Fvite%40063d93b',
  },
  {
    startup: 2515,
    serverStart: 201.99,
    fcp: 2045.5,
    index: 2,
    caseId: 'perf-2',
    displayName: '1000 React components',
    uniqueKey: 'sun0day%2Fvite%40999ad63',
  },
  {
    startup: 1990,
    serverStart: 166,
    fcp: 1578.3999998569489,
    index: 3,
    caseId: 'perf-1',
    displayName: 'vite 2.7 slow',
    uniqueKey: 'vitejs%2Fvite%40063d93b',
  },
  {
    startup: 1915,
    serverStart: 156.98,
    fcp: 1495.7999999523163,
    index: 3,
    caseId: 'perf-1',
    displayName: 'vite 2.7 slow',
    uniqueKey: 'sun0day%2Fvite%40999ad63',
  },
  {
    startup: 2484,
    serverStart: 199,
    fcp: 2026.7000000476837,
    index: 3,
    caseId: 'perf-2',
    displayName: '1000 React components',
    uniqueKey: 'vitejs%2Fvite%40063d93b',
  },
  {
    startup: 2575,
    serverStart: 199.65,
    fcp: 2101.5999999046326,
    index: 3,
    caseId: 'perf-2',
    displayName: '1000 React components',
    uniqueKey: 'sun0day%2Fvite%40999ad63',
  },
  {
    startup: 1991,
    serverStart: 163,
    fcp: 1573.5,
    index: 4,
    caseId: 'perf-1',
    displayName: 'vite 2.7 slow',
    uniqueKey: 'vitejs%2Fvite%40063d93b',
  },
  {
    startup: 1932,
    serverStart: 156.37,
    fcp: 1510,
    index: 4,
    caseId: 'perf-1',
    displayName: 'vite 2.7 slow',
    uniqueKey: 'sun0day%2Fvite%40999ad63',
  },
  {
    startup: 2600,
    serverStart: 209,
    fcp: 2117.199999809265,
    index: 4,
    caseId: 'perf-2',
    displayName: '1000 React components',
    uniqueKey: 'vitejs%2Fvite%40063d93b',
  },
  {
    startup: 2583,
    serverStart: 202.18,
    fcp: 2106.2999999523163,
    index: 4,
    caseId: 'perf-2',
    displayName: '1000 React components',
    uniqueKey: 'sun0day%2Fvite%40999ad63',
  },
]

const compares = [
  {
    owner: 'vitejs',
    repo: 'vite',
    ref: null,
    sha: '063d93bf5ed487bf89b74526d838711ed5e125eb',
    uniqueKey: 'vitejs%2Fvite%40063d93b',
  },
  {
    owner: 'sun0day',
    repo: 'vite',
    ref: null,
    sha: '999ad63afc4271aa37a960fba9d0aef4132efe61',
    uniqueKey: 'sun0day%2Fvite%40999ad63',
  },
]

const baseCases = [
  {
    id: 'perf-1',
    port: 5173,
    script: 'dev',
    displayName: 'vite 2.7 slow',
    viteCache: './node_modules/.vite',
  },
  {
    id: 'perf-2',
    port: 5173,
    script: 'start:vite',
    displayName: '1000 React components',
    viteCache: './node_modules/.vite',
  },
]
