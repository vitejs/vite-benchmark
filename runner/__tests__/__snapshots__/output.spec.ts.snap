// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`data report > composeGitHubActionsSummary 1`] = `
"# Benchmark for pull request [#12787](https://github.com/vitejs/vite/pull/12787)<h2>Meta Info</h2>
<table><tr><td>pull request link</td><td>vitejs/vite#12787</td></tr><tr><td>SHA of compare 0</td><td>vitejs/vite@063d93b</td></tr><tr><td>SHA of compare 1</td><td>sun0day/vite@999ad63</td></tr><tr><td>repetition</td><td>undefined</td></tr></table>
<h2>Benchmark Result</h2>
<h3>Case 1: <a href=\\"https://github.com/vitejs/vite-benchmark/tree/main/cases/perf-1\\">vite 2.7 slow</a></h3>
<table><tr><th>ref</th><th>start up mean</th><th>start up median</th><th>server start mean</th><th>server start median</th><th>fcp mean</th><th>fcp median</th></tr><tr><td>vitejs/vite@063d93b</td><td>2156</td><td>2000</td><td>223</td><td>166</td><td>1590</td><td>1578</td></tr><tr><td>sun0day/vite@999ad63</td><td>2078</td><td>1956</td><td>219</td><td>157</td><td>1522</td><td>1510</td></tr></table>
<table><tr><th>ref</th><th>start up mean</th><th>start up median</th><th>server start mean</th><th>server start median</th><th>fcp mean</th><th>fcp median</th></tr><tr><td>vitejs/vite@063d93b</td><td>-</td><td>-</td><td>-</td><td>-</td><td>-</td><td>-</td></tr><tr><td>sun0day/vite@999ad63</td><td>-78 (-3.62%) ‚ö°Ô∏è</td><td>-44 (-2.20%) ‚ö°Ô∏è</td><td>-4 (-1.79%) ‚ö°Ô∏è</td><td>-9 (-5.42%) ‚ö°Ô∏è</td><td>-68 (-4.28%) ‚ö°Ô∏è</td><td>-68 (-4.31%) ‚ö°Ô∏è</td></tr></table>
<h3>Case 2: <a href=\\"https://github.com/vitejs/vite-benchmark/tree/main/cases/perf-2\\">1000 React components</a></h3>
<table><tr><th>ref</th><th>start up mean</th><th>start up median</th><th>server start mean</th><th>server start median</th><th>fcp mean</th><th>fcp median</th></tr><tr><td>vitejs/vite@063d93b</td><td>2597</td><td>2600</td><td>202</td><td>203</td><td>2101</td><td>2089</td></tr><tr><td>sun0day/vite@999ad63</td><td>2592</td><td>2583</td><td>203</td><td>202</td><td>2092</td><td>2102</td></tr></table>
<table><tr><th>ref</th><th>start up mean</th><th>start up median</th><th>server start mean</th><th>server start median</th><th>fcp mean</th><th>fcp median</th></tr><tr><td>vitejs/vite@063d93b</td><td>-</td><td>-</td><td>-</td><td>-</td><td>-</td><td>-</td></tr><tr><td>sun0day/vite@999ad63</td><td>-5 (-0.19%) ‚ö°Ô∏è</td><td>-17 (-0.65%) ‚ö°Ô∏è</td><td>1 (+0.50%) üî∫</td><td>-1 (-0.49%) ‚ö°Ô∏è</td><td>-9 (-0.43%) ‚ö°Ô∏è</td><td>13 (+0.62%) üî∫</td></tr></table>
"
`;

exports[`data report > composeSummarized 1`] = `
{
  "perf-1": [
    {
      "caseId": "perf-1",
      "displayName": "vite 2.7 slow",
      "metrics": {
        "fcpStat": {
          "mean": "1590",
          "median": "1578",
        },
        "serverStartStat": {
          "mean": "223",
          "median": "166",
        },
        "startupStat": {
          "mean": "2156",
          "median": "2000",
        },
      },
      "repoRef": "vitejs/vite@063d93b",
    },
    {
      "caseId": "perf-1",
      "displayName": "vite 2.7 slow",
      "metrics": {
        "fcpStat": {
          "mean": "1522",
          "median": "1510",
        },
        "serverStartStat": {
          "mean": "219",
          "median": "157",
        },
        "startupStat": {
          "mean": "2078",
          "median": "1956",
        },
      },
      "repoRef": "sun0day/vite@999ad63",
    },
  ],
  "perf-2": [
    {
      "caseId": "perf-2",
      "displayName": "1000 React components",
      "metrics": {
        "fcpStat": {
          "mean": "2101",
          "median": "2089",
        },
        "serverStartStat": {
          "mean": "202",
          "median": "203",
        },
        "startupStat": {
          "mean": "2597",
          "median": "2600",
        },
      },
      "repoRef": "vitejs/vite@063d93b",
    },
    {
      "caseId": "perf-2",
      "displayName": "1000 React components",
      "metrics": {
        "fcpStat": {
          "mean": "2092",
          "median": "2102",
        },
        "serverStartStat": {
          "mean": "203",
          "median": "202",
        },
        "startupStat": {
          "mean": "2592",
          "median": "2583",
        },
      },
      "repoRef": "sun0day/vite@999ad63",
    },
  ],
}
`;
